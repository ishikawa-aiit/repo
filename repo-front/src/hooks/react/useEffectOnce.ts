import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
