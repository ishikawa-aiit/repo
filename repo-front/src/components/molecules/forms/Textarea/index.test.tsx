import { render } from '@testing-library/react'

import { Textarea } from '.'

describe('snapshots', () => {
  it('最新のスナップショットに更新されている', () => {
    const { container } = render(<Textarea />)

    expect(container).toMatchSnapshot()
  })
})
