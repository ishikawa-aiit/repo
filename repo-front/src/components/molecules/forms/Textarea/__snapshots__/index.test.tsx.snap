// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 最新のスナップショットに更新されている 1`] = `
.emotion-0 {
  overflow: hidden;
  resize: none;
}

<div>
  <textarea
    class="chakra-textarea emotion-0"
  />
</div>
`;
